File: crates/store-db/src/items.rs
//! Catalog reads and writes.
//!
//! The lookup path here is the one an operator feels: §11 budgets `<100 ms` for
//! resolving a scan. It is a single indexed query against `item_barcodes` and
//! `items`, joined to the cached `item_stock` — never a sum over the ledger.
//! That cache is exactly what §7 buys us.

use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use store_core::alert::AlertLevel;
use store_core::item::{validate_barcode, validate_item_code, BarcodeKind};
use uuid::Uuid;

use crate::error::{DbError, FoundExt, Result};

/// What the tablet needs after a scan: the item, where it lives, and how many
/// the system thinks are there.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ItemView {
    pub id: Uuid,
    pub item_code: String,
    pub description: String,
    pub uom: String,
    pub category_id: Option<Uuid>,
    pub category_name: Option<String>,
    pub iso_code: Option<String>,
    pub grade: Option<String>,
    pub manufacturer: Option<String>,
    pub mfr_part_no: Option<String>,
    pub bin_location: Option<String>,
    pub unit_cost: Option<Decimal>,
    pub reorder_level: Decimal,
    pub reorder_qty: Option<Decimal>,
    pub allow_negative: bool,
    pub active: bool,
    pub on_hand: Decimal,
    pub alert_state: String,
}

impl ItemView {
    pub fn alert_level(&self) -> AlertLevel {
        self.alert_state.parse().unwrap_or(AlertLevel::Ok)
    }
}

/// Fields accepted when creating or updating a catalog item.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ItemInput {
    pub item_code: String,
    pub description: String,
    pub category_id: Option<Uuid>,
    pub uom: String,
    pub iso_code: Option<String>,
    pub grade: Option<String>,
    pub manufacturer: Option<String>,
    pub mfr_part_no: Option<String>,
    pub diameter_mm: Option<Decimal>,
    pub flutes: Option<i32>,
    pub reorder_level: Decimal,
    pub reorder_qty: Option<Decimal>,
    pub bin_location: Option<String>,
    pub unit_cost: Option<Decimal>,
    pub allow_negative: bool,
}

/// Resolve a scanned barcode to an item.
///
/// Checks `items.item_code` first (our own labels, the common case) and then
/// `item_barcodes` (manufacturer and vendor labels). Returns `None` rather than
/// an error: an unknown barcode is an ordinary event on the shop floor, not a
/// failure.
pub async fn lookup_by_barcode(pool: &PgPool, raw_code: &str) -> Result<Option<ItemView>> {
    let code = validate_barcode(raw_code)?;

    let row = sqlx::query_as!(
        ItemView,
        r#"
        select i.id                as "id!",
               i.item_code         as "item_code!",
               i.description       as "description!",
               i.uom               as "uom!",
               i.category_id,
               c.name              as "category_name?",
               i.iso_code,
               i.grade,
               i.manufacturer,
               i.mfr_part_no,
               i.bin_location,
               i.unit_cost,
               i.reorder_level     as "reorder_level!",
               i.reorder_qty,
               i.allow_negative    as "allow_negative!",
               i.active            as "active!",
               coalesce(s.on_hand, 0)        as "on_hand!",
               coalesce(s.alert_state, 'OK') as "alert_state!"
          from items i
          left join item_categories c on c.id = i.category_id
          left join item_stock s      on s.item_id = i.id
         where i.item_code = $1
            or exists (select 1 from item_barcodes b
                        where b.item_id = i.id and b.code = $1)
         limit 1
        "#,
        code
    )
    .fetch_optional(pool)
    .await?;

    Ok(row)
}

/// Typeahead across item_code, description, iso_code and grade (§11).
///
/// Ranked so an exact or prefix match on the item code beats a description
/// fragment — an operator who types a code wants that code, not everything
/// mentioning it.
pub async fn search(pool: &PgPool, q: &str, limit: i64) -> Result<Vec<ItemView>> {
    let needle = q.trim();
    if needle.is_empty() {
        return Ok(Vec::new());
    }
    let pattern = format!("%{}%", needle.to_uppercase());
    let prefix = format!("{}%", needle.to_uppercase());

    let rows = sqlx::query_as!(
        ItemView,
        r#"
        select i.id                as "id!",
               i.item_code         as "item_code!",
               i.description       as "description!",
               i.uom               as "uom!",
               i.category_id,
               c.name              as "category_name?",
               i.iso_code,
               i.grade,
               i.manufacturer,
               i.mfr_part_no,
               i.bin_location,
               i.unit_cost,
               i.reorder_level     as "reorder_level!",
               i.reorder_qty,
               i.allow_negative    as "allow_negative!",
               i.active            as "active!",
               coalesce(s.on_hand, 0)        as "on_hand!",
               coalesce(s.alert_state, 'OK') as "alert_state!"
          from items i
          left join item_categories c on c.id = i.category_id
          left join item_stock s      on s.item_id = i.id
         where i.active
           and (upper(i.item_code) like $1
             or upper(i.description) like $1
             or upper(i.iso_code) like $1
             or upper(i.grade) like $1)
         order by case
                    when upper(i.item_code) = upper($3) then 0
                    when upper(i.item_code) like $2 then 1
                    when upper(i.iso_code) like $2 then 2
                    else 3
                  end,
                  i.item_code
         limit $4
        "#,
        pattern,
        prefix,
        needle,
        limit.clamp(1, 100)
    )
    .fetch_all(pool)
    .await?;

    Ok(rows)
}

/// Fetch one item by id.
pub async fn get(pool: &PgPool, id: Uuid) -> Result<ItemView> {
    sqlx::query_as!(
        ItemView,
        r#"
        select i.id                as "id!",
               i.item_code         as "item_code!",
               i.description       as "description!",
               i.uom               as "uom!",
               i.category_id,
               c.name              as "category_name?",
               i.iso_code,
               i.grade,
               i.manufacturer,
               i.mfr_part_no,
               i.bin_location,
               i.unit_cost,
               i.reorder_level     as "reorder_level!",
               i.reorder_qty,
               i.allow_negative    as "allow_negative!",
               i.active            as "active!",
               coalesce(s.on_hand, 0)        as "on_hand!",
               coalesce(s.alert_state, 'OK') as "alert_state!"
          from items i
          left join item_categories c on c.id = i.category_id
          left join item_stock s      on s.item_id = i.id
         where i.id = $1
        "#,
        id
    )
    .fetch_optional(pool)
    .await?
    .or_not_found("item")
}

/// Stock view filters (§11: `GET /api/v1/stock`).
#[derive(Debug, Clone, Default)]
pub struct StockFilter {
    /// Only items at or below their reorder level.
    pub low: bool,
    /// Only items at zero (or below).
    pub empty: bool,
    pub category_id: Option<Uuid>,
    pub bin_location: Option<String>,
    pub limit: i64,
    pub offset: i64,
}

/// The stock dashboard query.
pub async fn stock(pool: &PgPool, filter: &StockFilter) -> Result<Vec<ItemView>> {
    let rows = sqlx::query_as!(
        ItemView,
        r#"
        select i.id                as "id!",
               i.item_code         as "item_code!",
               i.description       as "description!",
               i.uom               as "uom!",
               i.category_id,
               c.name              as "category_name?",
               i.iso_code,
               i.grade,
               i.manufacturer,
               i.mfr_part_no,
               i.bin_location,
               i.unit_cost,
               i.reorder_level     as "reorder_level!",
               i.reorder_qty,
               i.allow_negative    as "allow_negative!",
               i.active            as "active!",
               s.on_hand           as "on_hand!",
               s.alert_state       as "alert_state!"
          from items i
          join item_stock s           on s.item_id = i.id
          left join item_categories c on c.id = i.category_id
         where i.active
           and (not $1::bool or s.alert_state in ('LOW', 'EMPTY'))
           and (not $2::bool or s.alert_state = 'EMPTY')
           and ($3::uuid is null or i.category_id = $3)
           and ($4::text is null or i.bin_location = $4)
         order by case s.alert_state when 'EMPTY' then 0 when 'LOW' then 1 else 2 end,
                  i.item_code
         limit $5 offset $6
        "#,
        filter.low,
        filter.empty,
        filter.category_id,
        filter.bin_location,
        filter.limit.clamp(1, 1000),
        filter.offset.max(0),
    )
    .fetch_all(pool)
    .await?;

    Ok(rows)
}

/// Create a catalog item.
///
/// The `item_code` is validated through `store-core` first, because the code is
/// also the Code128 payload printed on the bin label (§6) — a code we cannot
/// print is a code we must not accept.
pub async fn create(pool: &PgPool, input: &ItemInput) -> Result<Uuid> {
    let item_code = validate_item_code(&input.item_code)?;
    let uom: store_core::Uom = input.uom.parse()?;

    let id = sqlx::query_scalar!(
        r#"
        insert into items (
            item_code, description, category_id, uom,
            iso_code, grade, manufacturer, mfr_part_no, diameter_mm, flutes,
            reorder_level, reorder_qty, bin_location, unit_cost, allow_negative
        )
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15)
        returning id
        "#,
        item_code,
        input.description.trim(),
        input.category_id,
        uom.as_str(),
        input.iso_code.as_deref(),
        input.grade.as_deref(),
        input.manufacturer.as_deref(),
        input.mfr_part_no.as_deref(),
        input.diameter_mm,
        input.flutes,
        input.reorder_level,
        input.reorder_qty,
        input.bin_location.as_deref(),
        input.unit_cost,
        input.allow_negative,
    )
    .fetch_one(pool)
    .await
    .map_err(|e| DbError::from_sqlx(e, "item code"))?;

    Ok(id)
}

/// Update a catalog item in place.
///
/// Note that nothing here touches stock. Changing `reorder_level` re-evaluates
/// the alert state, but that happens in a trigger (`0003_ledger.up.sql`), not
/// here — one rule, one place.
pub async fn update(pool: &PgPool, id: Uuid, input: &ItemInput) -> Result<()> {
    let item_code = validate_item_code(&input.item_code)?;
    let uom: store_core::Uom = input.uom.parse()?;

    let result = sqlx::query!(
        r#"
        update items
           set item_code = $2, description = $3, category_id = $4, uom = $5,
               iso_code = $6, grade = $7, manufacturer = $8, mfr_part_no = $9,
               diameter_mm = $10, flutes = $11,
               reorder_level = $12, reorder_qty = $13, bin_location = $14,
               unit_cost = $15, allow_negative = $16
         where id = $1
        "#,
        id,
        item_code,
        input.description.trim(),
        input.category_id,
        uom.as_str(),
        input.iso_code.as_deref(),
        input.grade.as_deref(),
        input.manufacturer.as_deref(),
        input.mfr_part_no.as_deref(),
        input.diameter_mm,
        input.flutes,
        input.reorder_level,
        input.reorder_qty,
        input.bin_location.as_deref(),
        input.unit_cost,
        input.allow_negative,
    )
    .execute(pool)
    .await
    .map_err(|e| DbError::from_sqlx(e, "item code"))?;

    crate::expect_one(result, "item")
}

/// Retire an item without deleting it. Deleting would orphan ledger history.
pub async fn deactivate(pool: &PgPool, id: Uuid) -> Result<()> {
    let result = sqlx::query!("update items set active = false where id = $1", id)
        .execute(pool)
        .await?;
    crate::expect_one(result, "item")
}

/// Attach an alternate barcode (a manufacturer's EAN, a distributor's label).
pub async fn add_barcode(
    pool: &PgPool,
    item_id: Uuid,
    raw_code: &str,
    kind: BarcodeKind,
) -> Result<Uuid> {
    let code = validate_barcode(raw_code)?;

    let id = sqlx::query_scalar!(
        "insert into item_barcodes (item_id, code, kind) values ($1, $2, $3) returning id",
        item_id,
        code,
        kind.as_str(),
    )
    .fetch_one(pool)
    .await
    .map_err(|e| DbError::from_sqlx(e, "barcode"))?;

    Ok(id)
}

/// Items due for a label print run (§11: `POST /api/v1/admin/labels/print`).
pub async fn for_label_batch(pool: &PgPool, ids: &[Uuid]) -> Result<Vec<ItemView>> {
    let rows = sqlx::query_as!(
        ItemView,
        r#"
        select i.id                as "id!",
               i.item_code         as "item_code!",
               i.description       as "description!",
               i.uom               as "uom!",
               i.category_id,
               c.name              as "category_name?",
               i.iso_code,
               i.grade,
               i.manufacturer,
               i.mfr_part_no,
               i.bin_location,
               i.unit_cost,
               i.reorder_level     as "reorder_level!",
               i.reorder_qty,
               i.allow_negative    as "allow_negative!",
               i.active            as "active!",
               coalesce(s.on_hand, 0)        as "on_hand!",
               coalesce(s.alert_state, 'OK') as "alert_state!"
          from items i
          left join item_categories c on c.id = i.category_id
          left join item_stock s      on s.item_id = i.id
         where i.id = any($1)
         order by i.bin_location nulls last, i.item_code
        "#,
        ids
    )
    .fetch_all(pool)
    .await?;

    Ok(rows)
}

File: cloud/src/lib/items.ts
// Catalog reads (CLAUDE.md §11, §12.4).
//
// §12.4: the scan path and the search path must land on the same item card.
// They do, because both come back through `ItemRow` and both join the same
// stock read model.

import { sql } from "./db.ts";
import { ApiError } from "./api-error.ts";
import type { AlertState } from "./ledger.ts";

export interface ItemRow {
  id: string;
  item_code: string;
  description: string;
  uom: string;
  category_id: string | null;
  category_name: string | null;
  iso_code: string | null;
  grade: string | null;
  manufacturer: string | null;
  mfr_part_no: string | null;
  bin_location: string | null;
  unit_cost: string | null;
  reorder_level: string;
  reorder_qty: string | null;
  allow_negative: boolean;
  active: boolean;
  on_hand: string;
  alert_state: AlertState;
  last_txn_at: Date | null;
}

/**
 * The shared projection.
 *
 * A function, not a constant: a tagged template is a *call*, so evaluating it
 * at module scope opens a connection at import time — which defeats the
 * lazy client in `db.ts` and fails `next build`, where no DATABASE_URL exists
 * and none is needed. Local builds hid it behind a populated .env.local; CI
 * did not.
 */
const select = () => sql`
  select i.id, i.item_code, i.description, i.uom,
         i.category_id, c.name as category_name,
         i.iso_code, i.grade, i.manufacturer, i.mfr_part_no,
         i.bin_location, i.unit_cost::text as unit_cost,
         i.reorder_level::text as reorder_level,
         i.reorder_qty::text as reorder_qty,
         i.allow_negative, i.active,
         coalesce(s.on_hand, 0)::text as on_hand,
         coalesce(s.alert_state, 'OK') as alert_state,
         s.last_txn_at
    from items i
    left join item_categories c on c.id = i.category_id
    left join item_stock s on s.item_id = i.id
`;

export async function getItem(id: string): Promise<ItemRow> {
  const rows = await sql<ItemRow[]>`${select()} where i.id = ${id} limit 1`;
  const item = rows[0];
  if (!item) throw ApiError.notFound("no such item");
  return item;
}

/**
 * Resolve a scanned code to an item.
 *
 * Three things can be under the camera and all of them are legitimate:
 *
 *   our own item code        printed on the bin label
 *   an alternate barcode     the vendor's own printed EAN, so a box that
 *                            arrives labelled resolves without re-labelling
 *   a tool serial            the running QR number on an individual tool
 *
 * Order matters only for speed; the three namespaces are disjoint by unique
 * index, so a code cannot mean two items.
 */
export async function lookupByCode(code: string): Promise<ItemRow> {
  const trimmed = code.trim();
  if (!trimmed) throw ApiError.badRequest("empty barcode");

  const rows = await sql<ItemRow[]>`
    ${select()}
    where i.item_code = ${trimmed}
       or i.id = (select item_id from item_barcodes where code = ${trimmed})
       or i.id = (select item_id from tool_serials where serial_no = ${trimmed})
    limit 1
  `;

  const item = rows[0];
  if (!item) throw ApiError.notFound(`Nothing in the catalog matches ${trimmed}.`);
  return item;
}

/**
 * Typeahead across item_code, description, iso_code and grade (§11).
 *
 * Trigram matching because operators search by fragment — "cnmg 1204", "tn20" —
 * not by prefix. The GIN indexes in 0001 are what keep this inside the 100 ms
 * §11 asks for.
 */
export async function searchItems(query: string, limit = 25): Promise<ItemRow[]> {
  const q = query.trim();
  if (!q) return [];

  return sql<ItemRow[]>`
    ${select()}
    where i.active
      and (i.item_code ilike ${"%" + q + "%"}
        or i.description ilike ${"%" + q + "%"}
        or i.iso_code ilike ${"%" + q + "%"}
        or i.grade ilike ${"%" + q + "%"})
    order by
      -- An exact code match is what a scanner-shaped query wants first.
      (i.item_code ilike ${q}) desc,
      (i.item_code ilike ${q + "%"}) desc,
      i.item_code
    limit ${Math.min(Math.max(limit, 1), 100)}
  `;
}

/** The browse-all list: paged, stable order, active items only. */
export async function browseItems(offset = 0, limit = 25): Promise<ItemRow[]> {
  return sql<ItemRow[]>`
    ${select()}
    where i.active
    order by i.item_code
    limit ${Math.min(Math.max(limit, 1), 100)}
    offset ${Math.max(offset, 0)}
  `;
}

/**
 * The stock view, and also the admin console's catalog list.
 *
 * It returns the *whole* item, not just the stock-shaped columns. The console
 * loads this list and then opens the edit form straight from the row it has —
 * so a field missing here is a field the form opens blank and saves blank.
 * `active` is the same story: absent, every row renders as retired.
 */
export async function stockList(filters: {
  states?: string[] | null;
  q?: string | null;
  bin?: string | null;
  category?: string | null;
  limit?: number;
  /** Rows to skip. Absent means the first page. */
  offset?: number;
}): Promise<ItemRow[]> {
  const { states = null, q = null, bin = null, category = null, limit = 500, offset = 0 } =
    filters;

  return sql<ItemRow[]>`
    ${select()}
    where i.active
      and (${states}::text[] is null
           or coalesce(s.alert_state, 'OK') = any(${states}::text[]))
      and (${bin}::text is null or i.bin_location = ${bin})
      and (${category}::uuid is null or i.category_id = ${category}::uuid)
      and (${q}::text is null
           or i.item_code ilike '%' || ${q} || '%'
           or i.description ilike '%' || ${q} || '%')
    order by (coalesce(s.alert_state,'OK') = 'EMPTY') desc,
             (coalesce(s.alert_state,'OK') = 'LOW') desc,
             i.item_code
    limit ${Math.min(Math.max(limit, 1), 500)}
    offset ${Math.max(offset, 0)}
  `;
}
